### PR TITLE
Command/Inst refactor: Pulses and Play

### PR DESCRIPTION
**Pull request by lcapelluto**
### Summary

Part 6/7 for issue #3750

Migrate commands to instructions, migrate waveforms to `pulse_lib`

### Details and comments

Deprecate PulseCommands, create new Play Instruction-implementation. The Play instruction takes a Pulse and a PulseChannel. Pulse is a new class within the pulse library that is very close to the deprecated PulseCommand abstract class. The PulseCommand implementations (SamplePulse and the ParametricPulse subclasses) have been moved to the pulse library as well, subclassing from Pulse, and all valid arguments to Play.

TODO
- [x] pulses also need to have `names`
- [x] `assemble_schedules` needs to handle `Play`
- [x] `pulse_instruction` needs to handle `Play`
- [x] Playtests

### CODE REVIEW

File: qiskit/pulse/pulse_lib/pulse.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+from qiskit.pulse.channels import PulseChannel
+from .instruction import Instruction
+from .command import Command
+
+
+class Pulse(ABC):
+    """The abstract superclass for pulses."""
+
+    @abstractmethod  #???
+    def __init__(self, duration: int = None):
```
Comment (nkanazawa1989) [inline qiskit/pulse/pulse_lib/pulse.py]: why it doesn't have a `name`?

Comment (lcapelluto) [reply qiskit/pulse/pulse_lib/pulse.py]: because it is a work in progress

Comment (lcapelluto) [reply qiskit/pulse/pulse_lib/pulse.py]: it's not ready for review yet

File: qiskit/pulse/pulse_lib/parametric_pulses.py [hunk 2]
```diff
@@ -0,0 +1,399 @@
...
+    @abstractmethod
+    def __init__(self, duration: int, name: Optional[str] = None):
+        """Create a parametric pulse and validate the input parameters.
+
+        Args:
+            duration: Pulse length in terms of the the sampling period `dt`.
+            name: Display name for this pulse envelope.
+        """
+        super().__init__(duration=duration, name=name)
+        self.validate_parameters()
```
Comment (taalexander) [inline qiskit/pulse/pulse_lib/parametric_pulses.py]: Should validation be optional?

File: qiskit/pulse/pulse_lib/parametric_pulses.py [hunk 3]
```diff
@@ -0,0 +1,399 @@
...
+from .discrete import gaussian, gaussian_square, drag, constant
+from .pulse import Pulse
+from .sample_pulse import SamplePulse
+
+
+class ParametricPulse(Pulse):
+    """The abstract superclass for parametric pulses."""
+
+    @abstractmethod
+    def __init__(self, duration: int, name: Optional[str] = None):
```
Comment (taalexander) [inline qiskit/pulse/pulse_lib/parametric_pulses.py]: Should `parameters` be a standard input to the pulse?

File: qiskit/pulse/pulse_lib/parametric_pulses.py [hunk 4]
```diff
@@ -0,0 +1,399 @@
...
+        if abs(self.amp) > 1.:
+            raise PulseError("The amplitude norm must be <= 1, "
+                             "found: {}".format(abs(self.amp)))
+        if self.sigma <= 0:
+            raise PulseError("Sigma must be greater than 0.")
+        if self.width < 0 or self.width >= self.duration:
+            raise PulseError("The pulse width must be at least 0 and less than its duration.")
+
+    @property
+    def parameters(self) -> Dict[str, Any]:
```
Comment (taalexander) [inline qiskit/pulse/pulse_lib/parametric_pulses.py]: It should be able to extract an ordered list of parameters. I worry this is relying on the implementation of the Python dictionary. I would have preferred `parameters` to return a list and `named_parameters` to have returned the dictionary.

File: qiskit/pulse/pulse_lib/parametric_pulses.py [hunk 5]
```diff
@@ -0,0 +1,399 @@
...
+
+    @abstractmethod
+    def validate_parameters(self) -> None:
+        """
+        Validate parameters.
+
+        Raises:
+            PulseError: If the parameters passed are not valid.
+        """
+        raise NotImplementedError
```
Comment (taalexander) [inline qiskit/pulse/pulse_lib/parametric_pulses.py]: Should this be required to be implemented, or just pass by default?

File: qiskit/pulse/pulse_lib/pulse.py [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+from abc import ABC, abstractmethod
+
+import numpy as np
+
+from ..channels import PulseChannel
+from ..exceptions import PulseError
+from ..instructions.play import Play
+
+
+class Pulse(ABC):
```
Comment (taalexander) [inline qiskit/pulse/pulse_lib/pulse.py]: This is more akin to a `Type`, do you think it should live in a `types` module along with Kernel and Discriminator?

File: qiskit/pulse/pulse_lib/parametric_pulses.py (hunk 3 above)
Comment (lcapelluto) [reply qiskit/pulse/pulse_lib/parametric_pulses.py]: What would they be?

File: qiskit/pulse/pulse_lib/parametric_pulses.py (hunk 5 above)
Comment (lcapelluto) [reply qiskit/pulse/pulse_lib/parametric_pulses.py]: Do you prefer it to pass by default?

File: qiskit/pulse/pulse_lib/parametric_pulses.py (hunk 2 above)
Comment (lcapelluto) [reply qiskit/pulse/pulse_lib/parametric_pulses.py]: I don't think the runtime implications here are enough to warrant not validating

File: qiskit/pulse/pulse_lib/pulse.py (hunk 6 above)
Comment (lcapelluto) [reply qiskit/pulse/pulse_lib/pulse.py]: That's an interesting idea, I definitely like the idea of keeping simple, basic, "type-like" objects together and out of the way. For now, I think this does make the most sense. Once `__call__` is removed (since it's sort of un-type-like) and this command->instruction thing is done, we should work on another organizational refactor, including this idea.

File: qiskit/pulse/pulse_lib/parametric_pulses.py (hunk 4 above)
Comment (nkanazawa1989) [reply qiskit/pulse/pulse_lib/parametric_pulses.py]: does `OrderedDict` + `.values()` work?

File: qiskit/pulse/pulse_lib/parametric_pulses.py [hunk 7]
```diff
@@ -0,0 +1,399 @@
...
+        Args:
+            duration: Pulse length in terms of the the sampling period `dt`.
+            amp: The amplitude of the Gaussian envelope.
+            sigma: A measure of how wide or narrow the Gaussian peak is; described mathematically
+                   in the class docstring.
+            name: Display name for this pulse envelope.
+        """
+        self._amp = complex(amp)
+        self._sigma = sigma
+        super().__init__(duration=duration)
```
Comment (nkanazawa1989) [inline qiskit/pulse/pulse_lib/parametric_pulses.py]: why they don't pass `name`?

File: qiskit/assembler/assemble_schedules.py [hunk 8]
```diff
@@ -170,15 +171,31 @@ def _assemble_instructions(
...
             if pulse_shape not in run_config.parametric_pulses:
                 # Convert to SamplePulse if the backend does not support it
                 instruction = PulseInstruction(instruction.command.get_sample_pulse(),
                                                instruction.channels[0],
                                                name=instruction.name)
 
-        if isinstance(instruction, PulseInstruction):
+        if isinstance(instruction, Play) and isinstance(instruction.pulse, SamplePulse):
+            name = instruction.pulse.name
+            if instruction.pulse != user_pulselib.get(name):
```
Comment (nkanazawa1989) [inline qiskit/assembler/assemble_schedules.py]: if we have two sample pulses with the identical samples and with different names, it seems like the assembler recognizes them as different pulses. is this intended? or do we combine them to save hardware resources?

Comment (lcapelluto) [reply qiskit/assembler/assemble_schedules.py]: that's a good point. we should come back to this -- I'm going to say for now that it is out of scope for this PR (these lines duplicate existing behavior)

File: qiskit/pulse/pulse_lib/parametric_pulses.py (hunk 7 above)
Comment (lcapelluto) [reply qiskit/pulse/pulse_lib/parametric_pulses.py]: it should!!! thank you!!

File: qiskit/pulse/pulse_lib/parametric_pulses.py (hunk 4 above)
Comment (lcapelluto) [reply qiskit/pulse/pulse_lib/parametric_pulses.py]: `self.parameters[k] for k in sorted(self.parameters)`

File: releasenotes/notes/unify-instructions-and-commands-aaa6d8724b8a29d3.yaml [hunk 9]
```diff
@@ -25,9 +27,21 @@ features:
...
+    There is now a :py:class:`~qiskit.pulse.instructions.Play` instruction
+    which takes a description of a pulse envelope and a channel. The pulse
+    description must subclass from :py:class:`~qiskit.pulse.pulse_lib.Pulse`.
+    py:class:`~qiskit.pulse.pulse_lib.SamplePulse` and
+    py:class:`~qiskit.pulse.pulse_lib.ParametricPulse` s (e.g. ``Gaussian``)
+    now subclass from ``Pulse`` and have been moved to the ``pulse_lib``.
 
+    For example::
+
+        Play(SamplePulse[0.1]*10, DriveChannel(0))
```
Comment (eggerdj) [inline releasenotes/notes/unify-instructions-and-commands-aaa6d8724b8a29d3.yaml]:
```suggestion
        Play(SamplePulse([0.1]*10), DriveChannel(0))
```
?

File: releasenotes/notes/unify-instructions-and-commands-aaa6d8724b8a29d3.yaml [hunk 10]
```diff
@@ -25,9 +27,21 @@ features:
 
         sched += Delay(5, DriveChannel(0))
         sched += ShiftPhase(np.pi, DriveChannel(0))
-        sched += Acquire(100, AcquireChannel(0), MemorySlot(0))
+        sched += Play(SamplePulse([1.0, ...], DriveChannel(0))
```
Comment (eggerdj) [inline releasenotes/notes/unify-instructions-and-commands-aaa6d8724b8a29d3.yaml]:
```suggestion
        sched += Play(SamplePulse([1.0, ...]), DriveChannel(0))
```
?

File: test/python/qobj/test_pulse_converter.py [hunk 11]
```diff
@@ -207,13 +207,12 @@ def setUp(self):
     def test_drive_instruction(self):
         """Test converted qobj from PulseInstruction."""
         cmd = self.linear
```
Comment (eggerdj) [inline test/python/qobj/test_pulse_converter.py]:
```suggestion
```

File: test/python/qobj/test_pulse_converter.py [hunk 12]
```diff
@@ -207,13 +207,12 @@ def setUp(self):
...
-        instruction = cmd(DriveChannel(0)) << 10
+        instruction = cmd(DriveChannel(0))
```
Comment (eggerdj) [inline test/python/qobj/test_pulse_converter.py]:
```suggestion
        instruction = self.linear(DriveChannel(0))
```
why not?
There also seems to be other locations in this file where SamplePulse instances are named `command` see e.g. `test_drive_instruction`. Given the changes `pulse` might be a better name.
Can you go over this file and check if there are other things that need to be updated to reflect the current methodology?

File: qiskit/pulse/instructions/play.py [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+    """
+
+    def __init__(self, pulse: 'Pulse', channel: PulseChannel, name: Optional[str] = None):
+        """Create a new pulse instruction.
+
+        Args:
+            pulse: A pulse waveform description, such as
+                   :py:class:`~qiskit.pulse.pulse_lib.SamplePulse`.
+            channel: The channel that will have the pulse.
+            name: Name of the delay for display purposes.
```
[the remark this answers is not in the document]
Comment (lcapelluto) [reply qiskit/pulse/instructions/play.py]: Nothing 😱 thank you

File: releasenotes/notes/unify-instructions-and-commands-aaa6d8724b8a29d3.yaml [hunk 14]
```diff
@@ -11,12 +11,14 @@ features:
...
         sched += Delay(5)(DriveChannel(0))
         sched += ShiftPhase(np.pi)(DriveChannel(0))
+        sched += SamplePulse([1.0, ...])(DriveChannel(0))
         sched += Acquire(100)(AcquireChannel(0), MemorySlot(0))
 
     or, equivalently (though less used)::
 
         sched += DelayInstruction(Delay(5), DriveChannel(0))
         sched += ShiftPhaseInstruction(ShiftPhase(np.pi), DriveChannel(0))
+        sched += PulseInstruction(SamplePulse([1.0, ...], DriveChannel(0))
```
Comment (eggerdj) [inline releasenotes/notes/unify-instructions-and-commands-aaa6d8724b8a29d3.yaml]:
```suggestion
        sched += PulseInstruction(SamplePulse([1.0, ...]), DriveChannel(0))
```

Comment (lcapelluto) [reply releasenotes/notes/unify-instructions-and-commands-aaa6d8724b8a29d3.yaml]: thanks so much for catching these 😬

**Comment by lcapelluto** [comment]
@eggerdj @nkanazawa1989 thank you guys, I will try to fix plotting on this PR before merge

**Comment by lcapelluto** [comment]
Seems to fail on and off because sometimes Qobj `to_dict` doesn't return a dict. Very weird nondeterministic behavior that I couldn't reproduce locally

File: qiskit/visualization/pulse/matplotlib.py [hunk 15]
```diff
@@ -59,18 +59,20 @@ def __init__(self, t0, tf):
...
 
         Args:
             start_time (int): Starting time of instruction
-            pulse (Instruction): Instruction object to be added
+            instruction (Instruction): Instruction object to be added
         """
-
+        if isinstance(instruction, Delay):
+            return
+        pulse = getattr(instruction, 'pulse', instruction.command)
```
Comment (nkanazawa1989) [inline qiskit/visualization/pulse/matplotlib.py]: `ShiftPhase` and `SetFrequency` don't have `.command` and this seems to return `None`. This is how previous bug was caused.

Comment (lcapelluto) [reply qiskit/visualization/pulse/matplotlib.py]: Ah, I'm really unfamiliar with this, thanks. from the name of the function and the argument names it seemed like this wouldn't process non pulse instructions. I'll update

Comment (lcapelluto) [reply qiskit/visualization/pulse/matplotlib.py]: I'm wrong -- the name of the function should have made it clear

**Comment by lcapelluto** [comment]
this is getting messy. Going to make some updates today and then reopen when it's closer

**Comment by lcapelluto** [comment]
my mistake

File: releasenotes/notes/unify-instructions-and-commands-aaa6d8724b8a29d3.yaml [hunk 16]
```diff
@@ -56,6 +70,12 @@ deprecations:
     calling a command on a channel will be supported::
 
         ShiftPhase(<phase>)(<channel>)
+  - |
+    The call method of py:class:`~qiskit.pulse.pulse_lib.SamplePulse` and
```
Comment (mtreinish) [inline releasenotes/notes/unify-instructions-and-commands-aaa6d8724b8a29d3.yaml]:
```suggestion
    The ``call`` method of py:class:`~qiskit.pulse.pulse_lib.SamplePulse` and
```

File: releasenotes/notes/unify-instructions-and-commands-aaa6d8724b8a29d3.yaml [hunk 17]
```diff
@@ -25,9 +27,21 @@ features:
...
+        sched += Play(SamplePulse([1.0, ...]), DriveChannel(0))
         sched += SetFrequency(5.5, DriveChannel(0))  # New instruction!
+        sched += Acquire(100, AcquireChannel(0), MemorySlot(0))
+  - |
+    There is now a :py:class:`~qiskit.pulse.instructions.Play` instruction
+    which takes a description of a pulse envelope and a channel. The pulse
+    description must subclass from :py:class:`~qiskit.pulse.pulse_lib.Pulse`.
+    py:class:`~qiskit.pulse.pulse_lib.SamplePulse` and
+    py:class:`~qiskit.pulse.pulse_lib.ParametricPulse` s (e.g. ``Gaussian``)
+    now subclass from ``Pulse`` and have been moved to the ``pulse_lib``.
```
Comment (mtreinish) [inline releasenotes/notes/unify-instructions-and-commands-aaa6d8724b8a29d3.yaml]: This was a bit hard to follow, Originally I thought this should have been the pulse description should either be `Pulse`, `SamplePulse`, or `ParametricPulse`. But then I realized what it was actually saying. Maybe we should split this into 2 release notes. One saying that `SamplePulse` and `ParametricPulse` are now subclasses of `Pulse` and have been moved to `pulse_lib` (maybe as an upgrade note?) and the other about `Play`.

Comment (lcapelluto) [reply releasenotes/notes/unify-instructions-and-commands-aaa6d8724b8a29d3.yaml]: This is a good idea, thank you!
